### ISSUE TITLE
Make Concent refuse Clients with the same Ethereum keys

### ISSUE THREAD

**Issue by rwrzesien** [golemfactory/concent#943]
According to @shadeofblue comment, there should not be two clients using the same Ethereum key. If so, Concent should recognize this situation and act properly.

Changes in models:
1. Modify `ClientManager.get_or_create_full_clean` to require Ethereum key. Then, it should have following flow:

1a. If the `Client` with given `public_key` and `DepositAccount` both do not exist, create new `Client`.
1b. If the `Client` with given `public_key` and `DepositAccount` with given `ethereum_key` both exist, and are related, return matched `Client`.
1c. If the `Client` with given `public_key` exists but `DepositAccount` do not, return matched `Client`.
1d. If the `Client` with given `public_key` and `DepositAccount` with given `ethereum_key` both exist, and are **not** related, raise `ValidationError`.
1e. If the `Client` with given `public_key` does not exist but `DepositAccount` exists, raise `ValidationError`.

Changes in Bankster:
1. In `finalize_payment`, when the use case is `ADDITIONAL_VERIFICATION`, decide if payment should be done for requestor or provided basing on both `Client` and `payer_deposit_account.ethereum_address` instead of only `payer_deposit_account.ethereum_address`.

Tests:
1. Add unit/django-integration tests that Concent will refuse to process message with `TaskToCompute` using already existing Ethereum key with different public key.
2. It would be good to test scenario described in https://github.com/golemfactory/concent/issues/943#issuecomment-444134059.
3. Update e2e tests to use either same combination of public/Ethereum keys for each `subtask_id` (this can be done as a separate issue/pull request, but has to be merged not after changes from this issue).

Update all related failing tests.
Update all related docs (especially Bankster blueprints).

**Comment by PaweuB** [comment]
Pivotal: 
https://www.pivotaltracker.com/story/show/162197343

**Comment by rwrzesien** [comment]
The problem occurred because in end-to-end tests for reach test we use predefined, harcoded ethereum keys, but different (generated) provider/requestor keys.

When code flow enters `claim_deposit` Bankster function, there is a call to `Client.get_or_create_full_clean`, which each time creates a new client because each time there is a different provider/requestor key. Much later this exact `Client` is assigned to `Subtask` object created at the end of use case handler.
Next, in Bankster, after call to `Client.get_or_create_full_clean`, there is a call to `DepositAccount.objects.get_or_create_full_clean` with parameters `client` (just created one) and `ethereum_address` (constant, hardcoded in tests). This method checks if there is already existing `DepositAccount`, but filtering only by `ethereum_address`. This results in returning `DepositAccount` object with different related `Client` assigned to it (which is created during first run on the empty database). That `DepositAccount` is then assigned to `DepositClaim`, so it caused mismatch before when we compared
```
subtask.provider.public_key == deposit_claim.payer_deposit_account.client.public_key or
subtask.requestor.public_key == deposit_claim.payer_deposit_account.client.public_key
```

As a confirmation, after adding this check in Bankster where it was previously (`finalize_payment` for ` ConcentUseCase.ADDITIONAL_VERIFICATION`), it works correctly for the first run on clean database.

We are waiting for decision from the Golem team if Concent should accept situation when two different clients use the same ethereum key.
If it does, we have to modify use case 5 and Bankster.
If it doesn't, we need to fix our tests and add validation against that to Concent.

**Comment by rwrzesien** [comment]
TL;DR: We are waiting for decision from the Golem team if Concent should accept situation when two different clients use the same ethereum key.

**Comment by shadeofblue** [comment]
@PaweuB @rwrzesien for now, we can assume that two different clients won't have two different ethereum keys ... but generally, in the long run, that assumption may prove incorrect

**Comment by rwrzesien** [comment]
Changed description on how to handle situation with two different clients trying to use the same Ethereum key.

**Comment by cameel** [comment]
> 3. Remove `DepositAccount` model.

Doing this would make it impossible to support one client having multiple Ethereum accounts. `DepositAccount` must stay.

**Comment by rwrzesien** [comment]
@cameel Right, I forgot about this. I have rewritten the proposed solution.

**Comment by dybi** [comment]
ok, so now we will need to generate the `requestor` / `provider` keys in our E2E tests? Is the mechanizm for that already implemented? @kbeker

**Comment by cameel** [comment]
#1025 is not implemented yet. As agreed yesterday, it's better to leave this task until we have #1025. Otherwise we'll lose the ability to run tests quickly in development.

**Comment by kbeker** [comment]
@dybi @cameel it is implemented. `SCIBaseTest` gives generated keys for `provider` / `requestor`, but to do it in "fly" to generate new keys every `api-e2e-test` you need to add `--full_sci` flag

**Comment by cameel** [comment]
@kbeker I don't see a pull request linked there and the issue is still open.

**Comment by dybi** [comment]
solved by `--full_sci` option - closing
